File: app/src/main/java/jp/hazuki/yuzubrowser/action/item/CustomSingleAction.java
package jp.hazuki.yuzubrowser.action.item;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Parcel;
import android.os.Parcelable;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonToken;

import java.io.IOException;

import jp.hazuki.yuzubrowser.action.Action;
import jp.hazuki.yuzubrowser.action.ActionNameArray;
import jp.hazuki.yuzubrowser.action.SingleAction;
import jp.hazuki.yuzubrowser.action.view.ActionActivity;
import jp.hazuki.yuzubrowser.utils.app.OnActivityResultListener;
import jp.hazuki.yuzubrowser.utils.app.StartActivityInfo;

public class CustomSingleAction extends SingleAction {
    private static final String FIELD_NAME_ACTION = "0";
    private static final String FIELD_NAME_ACTION_NAME = "1";
    private Action mAction;
    private String mName;

    public CustomSingleAction(int id, JsonParser parser) throws IOException {
        super(id);

        mAction = new Action();
        if (parser != null) {
            if (parser.nextToken() != JsonToken.START_OBJECT) return;
            while (parser.nextToken() != JsonToken.END_OBJECT) {
                if (parser.getCurrentToken() != JsonToken.FIELD_NAME) return;
                if (FIELD_NAME_ACTION.equals(parser.getCurrentName())) {
                    mAction.loadAction(parser);
                    continue;
                }
                if (FIELD_NAME_ACTION_NAME.equals(parser.getCurrentName())) {
                    if (parser.nextToken() != JsonToken.VALUE_STRING) return;
                    mName = parser.getText();
                    continue;
                }
                parser.skipChildren();
            }
        }
    }

    @Override
    public void writeIdAndData(JsonGenerator generator) throws IOException {
        generator.writeNumber(id);
        generator.writeStartObject();
        mAction.writeAction(FIELD_NAME_ACTION, generator);
        generator.writeStringField(FIELD_NAME_ACTION_NAME, mName);
        generator.writeEndObject();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(id);
        dest.writeParcelable(mAction, flags);
        dest.writeString(mName);
    }

    protected CustomSingleAction(Parcel source) {
        super(source.readInt());
        mAction = source.readParcelable(Action.class.getClassLoader());
        mName = source.readString();
    }

    public static final Parcelable.Creator<CustomSingleAction> CREATOR = new Parcelable.Creator<CustomSingleAction>() {
        @Override
        public CustomSingleAction createFromParcel(Parcel source) {
            return new CustomSingleAction(source);
        }

        @Override
        public CustomSingleAction[] newArray(int size) {
            return new CustomSingleAction[size];
        }
    };

    @Override
    public StartActivityInfo showMainPreference(ActionActivity context) {
        return showSubPreference(context);
    }

    @Override
    public StartActivityInfo showSubPreference(ActionActivity context) {
        Intent intent = new Intent(context, CustomSingleActionActivity.class);
        intent.putExtra(CustomSingleActionActivity.EXTRA_ACTION, (Parcelable) mAction);
        intent.putExtra(CustomSingleActionActivity.EXTRA_NAME, mName);
        intent.putExtra(ActionNameArray.INTENT_EXTRA, (Parcelable) context.getActionNameArray());

        OnActivityResultListener l = new OnActivityResultListener() {
            @Override
            public void onActivityResult(Context context, int resultCode, Intent intent) {
                if (resultCode != Activity.RESULT_OK || intent == null)
                    return;
                mAction = intent.getParcelableExtra(CustomSingleActionActivity.EXTRA_ACTION);
                mName = intent.getStringExtra(CustomSingleActionActivity.EXTRA_NAME);
            }
        };

        return new StartActivityInfo(intent, l);
    }

    public Action getAction() {
        return mAction;
    }

    @Override
    public String toString(ActionNameArray nameArray) {
        if (mName != null)
            return mName;
        return super.toString(nameArray);
    }
}

File: legacy/src/main/java/jp/hazuki/yuzubrowser/legacy/action/item/CustomSingleAction.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.legacy.action.item

import android.app.Activity
import android.content.Intent
import android.os.Parcel
import android.os.Parcelable
import com.squareup.moshi.JsonReader
import com.squareup.moshi.JsonWriter
import jp.hazuki.yuzubrowser.legacy.action.Action
import jp.hazuki.yuzubrowser.legacy.action.ActionNameArray
import jp.hazuki.yuzubrowser.legacy.action.SingleAction
import jp.hazuki.yuzubrowser.legacy.action.view.ActionActivity
import jp.hazuki.yuzubrowser.ui.app.StartActivityInfo
import java.io.IOException

class CustomSingleAction : SingleAction {
    var action: Action
        private set
    private var mName: String? = null

    @Throws(IOException::class)
    constructor(id: Int, reader: JsonReader?) : super(id) {
        action = Action()
        if (reader != null) {
            if (reader.peek() != JsonReader.Token.BEGIN_OBJECT) return
            reader.beginObject()
            while (reader.hasNext()) {
                if (reader.peek() != JsonReader.Token.NAME) return
                when (reader.nextName()) {
                    FIELD_NAME_ACTION -> {
                        action.loadAction(reader)
                    }
                    FIELD_NAME_ACTION_NAME -> {
                        if (reader.peek() == JsonReader.Token.NULL) {
                            reader.nextNull<String>()
                        } else {
                            if (reader.peek() == JsonReader.Token.STRING) {
                                mName = reader.nextString()
                            } else {
                                reader.skipValue()
                            }
                        }
                    }
                    else -> reader.skipValue()
                }
            }
            reader.endObject()
        }
    }

    @Throws(IOException::class)
    override fun writeIdAndData(writer: JsonWriter) {
        writer.value(id)
        writer.beginObject()
        writer.name(FIELD_NAME_ACTION)
        action.writeAction(writer)
        writer.name(FIELD_NAME_ACTION_NAME)
        writer.value(mName)
        writer.endObject()
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(id)
        dest.writeParcelable(action, flags)
        dest.writeString(mName)
    }

    private constructor(source: Parcel) : super(source.readInt()) {
        action = source.readParcelable(Action::class.java.classLoader)!!
        mName = source.readString()
    }

    override fun showMainPreference(context: ActionActivity): StartActivityInfo? {
        return showSubPreference(context)
    }

    override fun showSubPreference(context: ActionActivity): StartActivityInfo? {
        val intent = Intent(context, CustomSingleActionActivity::class.java)
        intent.putExtra(CustomSingleActionActivity.EXTRA_ACTION, action as Parcelable?)
        intent.putExtra(CustomSingleActionActivity.EXTRA_NAME, mName)
        intent.putExtra(ActionNameArray.INTENT_EXTRA, context.actionNameArray as Parcelable)

        return StartActivityInfo(intent) { _, resultCode, data ->
            if (resultCode != Activity.RESULT_OK || data == null)
                return@StartActivityInfo
            action = data.getParcelableExtra(CustomSingleActionActivity.EXTRA_ACTION)
            mName = data.getStringExtra(CustomSingleActionActivity.EXTRA_NAME)
        }
    }

    override fun toString(nameArray: ActionNameArray): String? {
        return if (mName != null) mName else super.toString(nameArray)
    }

    companion object {
        private const val FIELD_NAME_ACTION = "0"
        private const val FIELD_NAME_ACTION_NAME = "1"

        @JvmField
        val CREATOR: Parcelable.Creator<CustomSingleAction> = object : Parcelable.Creator<CustomSingleAction> {
            override fun createFromParcel(source: Parcel): CustomSingleAction {
                return CustomSingleAction(source)
            }

            override fun newArray(size: Int): Array<CustomSingleAction?> {
                return arrayOfNulls(size)
            }
        }
    }
}
